modules/javafx.graphics/src/main/java/com/sun/glass/ui/GlassRobot.java

```diff
@@ -0,0 +1,361 @@
+                for (int iy = 0; iy < height; iy++) {
+                    double rely = ((y + iy + 0.5f) * outputScaleY) - (pminy + 0.5f);
+                    int irely = (int) Math.floor(rely);
+                    int fracty = (int) ((rely - irely) * 256);
+                    for (int ix = 0; ix < width; ix++) {
+                        double relx = ((x + ix + 0.5f) * outputScaleX) - (pminx + 0.5f);
+                        int irelx = (int) Math.floor(relx);
+                        int fractx = (int) ((relx - irelx) * 256);
+                        data[index++] = interp(tmpdata, irelx, irely, pwidth, pheight, fractx, fracty);
+                    }
+                }
+                dw = (int) width;
+                dh = (int) height;
+            }
+        }
+
+        return convertFromPixels(image, Application.GetApplication().createPixels(dw, dh, IntBuffer.wrap(data)));
+    }
+
+    public static int convertToRobotMouseButton(MouseButton[] buttons) {
+        int ret = 0;
+        for (MouseButton button : buttons) {
+            switch (button) {
+                case PRIMARY: ret |= MOUSE_LEFT_BTN; break;
+                case SECONDARY: ret |= MOUSE_RIGHT_BTN; break;
+                case MIDDLE: ret |= MOUSE_MIDDLE_BTN; break;
+                default: throw new IllegalArgumentException("MouseButton: " + button + " not supported by Robot");
+            }
+        }
+        return ret;
+    }
+
+    public static Color convertFromIntArgb(int color) {
+        int alpha = (color >> 24) & 0xFF;
+        int red   = (color >> 16) & 0xFF;
+        int green = (color >>  8) & 0xFF;
+        int blue  =  color        & 0xFF;
+        return new Color(red / 255d, green / 255d, blue / 255d, alpha / 255d);
+    }
+
+    protected static WritableImage convertFromPixels(WritableImage image, Pixels pixels) {
+        Objects.requireNonNull(pixels);
+        int width = pixels.getWidth();
+        int height = pixels.getHeight();
+        if (image == null) {
+            image = new WritableImage(width, height);
+        }
+
+        int bytesPerComponent = pixels.getBytesPerComponent();
+        if (bytesPerComponent == 4) {
+            IntBuffer intBuffer = (IntBuffer) pixels.getPixels();
+            writeIntBufferToImage(intBuffer, image);
+        }
+        else if (bytesPerComponent == 1) {
+            ByteBuffer byteBuffer = (ByteBuffer) pixels.getPixels();
+            writeByteBufferToImage(byteBuffer, image);
+        }
+
+        return image;
+    }
+
+    private static void writeIntBufferToImage(IntBuffer intBuffer, WritableImage image) {
+        Objects.requireNonNull(image);
+        PixelWriter pixelWriter = image.getPixelWriter();
+        double width = image.getWidth();
+        double height = image.getHeight();
+
+        for (int y = 0; y < height; y++) {
+            for (int x = 0; x < width; x++) {
+                int argb = intBuffer.get();
+                pixelWriter.setArgb(x, y, argb);
```

Comment:
I recommend filing an RFE to use PixelWriter::setPixels rather than setArgb where possible for performance. It doesn't need to be done in this release, but we should track it.